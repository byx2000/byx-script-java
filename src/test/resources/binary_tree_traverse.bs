// 二叉树定义
function TreeNode(val, left, right) {
    return {val, left, right}
}

// 前序遍历
function preorderTraverse(root) {
    var result = []
    function doTraverse(root) {
        if (root == undefined) {
            return
        }
        result.addLast(root.val)
        doTraverse(root.left)
        doTraverse(root.right)
    }
    doTraverse(root)
    return result
}

// 中序遍历
function inorderTraverse(root) {
    var result = []
    function doTraverse(root) {
        if (root == undefined) {
            return
        }
        doTraverse(root.left)
        result.addLast(root.val)
        doTraverse(root.right)
    }
    doTraverse(root)
    return result
}

// 后序遍历
function postorderTraverse(root) {
    var result = []
    function doTraverse(root) {
        if (root == undefined) {
            return
        }
        doTraverse(root.left)
        doTraverse(root.right)
        result.addLast(root.val)
    }
    doTraverse(root)
    return result
}

// 层序遍历
function levelTraverse(root) {
    var result = []
    var queue = [root]
    while (!queue.isEmpty()) {
        var cnt = queue.length()
        for (var i = 0; i < cnt; ++i) {
            var n = queue.removeFirst()
            result.addLast(n.val)
            if (n.left != undefined) {
                queue.addLast(n.left)
            }
            if (n.right != undefined) {
                queue.addLast(n.right)
            }
        }
    }
    return result
}

var root = TreeNode(1, TreeNode(2, TreeNode(4, TreeNode(7), undefined), TreeNode(5)), TreeNode(3, undefined, TreeNode(6)))

Console.println(preorderTraverse(root))
Console.println(inorderTraverse(root))
Console.println(postorderTraverse(root))
Console.println(levelTraverse(root))
