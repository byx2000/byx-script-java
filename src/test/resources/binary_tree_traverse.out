[1, 2, 4, 7, 5, 3, 6]
[7, 4, 2, 5, 1, 3, 6]
[7, 4, 5, 2, 6, 3, 1]
[1, 2, 3, 4, 5, 6, 7]
